tools/wasi/wasi.cpp

```diff
@@ -0,0 +1,217 @@
+        static_cast<uvwasi_size_t>(iovs.size()), &nread);
+    uvwasi_serdes_write_uint32_t(instance.memory->data(), nread_ptr, nread);
+
+    return fizzy::Value{uint32_t{ret}};
+}
+
+fizzy::ExecutionResult wasi_fd_prestat_get(
+    fizzy::Instance& instance, fizzy::span<const fizzy::Value> args, int)
+{
+    const auto fd = args[0].as<uint32_t>();
+    const auto prestat_ptr = args[1].as<uint32_t>();
+
+    uvwasi_prestat_t buf;
+    uvwasi_errno_t ret = uvwasi_fd_prestat_get(&state, fd, &buf);
+
+    uvwasi_serdes_write_prestat_t(instance.memory->data(), prestat_ptr, &buf);
+
+    return fizzy::Value{uint32_t{ret}};
+}
+
+fizzy::ExecutionResult wasi_environ_sizes_get(
+    fizzy::Instance& instance, fizzy::span<const fizzy::Value> args, int)
+{
+    const auto environc = args[0].as<uint32_t>();
+    const auto environ_buf_size = args[1].as<uint32_t>();
+    // TODO: implement properly (only returns 0 now)
+    uvwasi_serdes_write_uint32_t(instance.memory->data(), environc, 0);
+    uvwasi_serdes_write_uint32_t(instance.memory->data(), environ_buf_size, 0);
+    return fizzy::Value{uint32_t{UVWASI_ESUCCESS}};
+}
+
+bool run(int argc, const char** argv)
+{
+    // TODO: support both wasi_snapshot_preview1 and wasi_unstable?
+    const std::vector<fizzy::ImportedFunction> wasi_functions = {
+        {"wasi_snapshot_preview1", "proc_exit", {fizzy::ValType::i32}, std::nullopt,
+            wasi_proc_exit},
+        {"wasi_snapshot_preview1", "fd_read",
+            {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32},
+            fizzy::ValType::i32, wasi_fd_read},
+        {"wasi_snapshot_preview1", "fd_write",
+            {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32},
+            fizzy::ValType::i32, wasi_fd_write},
+        {"wasi_snapshot_preview1", "fd_prestat_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+            fizzy::ValType::i32, wasi_fd_prestat_get},
+        {"wasi_snapshot_preview1", "fd_prestat_dir_name",
+            {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32}, fizzy::ValType::i32,
+            wasi_return_enosys},
+        {"wasi_snapshot_preview1", "environ_sizes_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+            fizzy::ValType::i32, wasi_environ_sizes_get},
+        {"wasi_snapshot_preview1", "environ_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+            fizzy::ValType::i32, wasi_return_enosys},
+    };
+
+    uvwasi_options_t options = {
+        3,           // sizeof fd_table
+        0, nullptr,  // NOTE: no remappings
+        static_cast<uvwasi_size_t>(argc), argv,
+        nullptr,  // TODO: support env
+        0, 1, 2,
+        nullptr,  // NOTE: no special allocator
+    };
+
+    uvwasi_errno_t err = uvwasi_init(&state, &options);
+    if (err != UVWASI_ESUCCESS)
+    {
+        std::cerr << "Failed to initialise UVWASI: " << uvwasi_embedder_err_code_to_string(err)
```

Comment:
How are you going to test this?

Reply:
Checking uvwasi this is triggered in two realistic cases (in version 0.10.0): 1) out of memory; 2) invalid paths supplied in `options->preopen`. We do not make use of that feature yet.